src/App.tsx

```diff
@@ -3,30 +3,172 @@ import './App.scss';
 import './styles/general.scss';
 import { TodoList } from './components/TodoList';
 import { CurrentUser } from './components/CurrentUser';
+import { getTodosFromServer } from './api';
 
 interface State {
-  selectedUserId: number;
+  todos: Todo[],
+  selectedUserId: number,
+  inputValue: string,
+  selectValue: string,
+  loading: boolean,
+  errorMessage: string,
 }
 
 class App extends React.Component<{}, State> {
   state: State = {
+    todos: [],
     selectedUserId: 0,
+    inputValue: '',
+    selectValue: 'all',
+    loading: false,
+    errorMessage: '',
+  };
+
+  componentDidMount() {
+    this.getTodos();
+  }
+
+  getTodos = async (completed?: boolean) => {
+    this.setState({
+      loading: true,
+    });
+
+    try {
+      const todosFromServer = await getTodosFromServer(completed);
+
+      this.setState({
+        todos: [...todosFromServer],
+      });
+    } catch (error) {
+      this.setState({
+        errorMessage: 'Oops... Server is not responding',
+      });
+    } finally {
+      setTimeout(() => {
+        this.setState({
+          loading: false,
+        });
+      }, 500);
+    }
+  };
+
+  selectUserHandler = (userId: string) => {
+    const userNewId = +userId;
+
+    if (this.state.selectedUserId !== userNewId) {
+      this.setState({ selectedUserId: userNewId });
+    }
+  };
+
+  changeInputValue = (event: React.ChangeEvent<HTMLInputElement>) => {
+    const { value } = event.target;
+
+    this.setState({
+      inputValue: value,
+    });
+  };
+
+  changeSelectValue = (event: React.ChangeEvent<HTMLSelectElement>) => {
+    const { value } = event.target;
+
+    this.filterBySelect(value);
+
+    this.setState({
+      selectValue: value,
+    });
+  };
+
+  filterBySelect = async (selectValue: string) => {
+    this.setState({
+      loading: true,
+    });
+
+    switch (selectValue) {
+      case 'completed':
+        this.getTodos(true);
+        break;
+
+      case 'not completed':
+        this.getTodos(false);
+        break;
+
+      default:
+        this.getTodos();
+        break;
+    }
+  };
+
+  getTodosFilteredByInput = (todos: Todo[]) => {
+    const { inputValue } = this.state;
+
+    if (inputValue.length) {
+      return todos
+        .filter(todo => todo.title.toLowerCase().includes(inputValue.toLowerCase()));
```

Comment:
create a separate variable for `inputValue.toLowerCase()`, no need to do this operation on every iteration